### PR TITLE
[JENKINS-35247] Moved GitStep to git-plugin

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-35247](https://issues.jenkins-ci.org/browse/JENKINS-35247)

Note that although this adds a new dependency, the transitive dependencies (intentionally) include no GUI extensions, so it should have no effect on non-Pipeline users.

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
Hmm, have a

    hudson.util.CyclicGraphDetector$CycleDetectedException: Cycle detected: Plugin:git -> Plugin:workflow-scm-step -> Plugin:git

which I need to figure out how to deal with.

File: src/main/resources/jenkins/plugins/git/GitStep/help.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<div>
```
Comment (KostyaSha) [inline src/main/resources/jenkins/plugins/git/GitStep/help.html]: AFAIR jenkins wraps all helps into divs and you don't need wrote it in help files.

File: src/test/java/jenkins/plugins/git/GitStepTest.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2014 Jesse Glick.
```
Comment (KostyaSha) [inline src/test/java/jenkins/plugins/git/GitStepTest.java]: 2014? :)

File: src/main/resources/jenkins/plugins/git/GitStep/help.html (hunk 1 above)
Comment (jglick) [inline src/main/resources/jenkins/plugins/git/GitStep/help.html]: Yeah, probably not necessary, but I like to have a root element in the file. Should be harmless anyway.

Note that this file is unmodified, it is just moved from another repository.

File: src/test/java/jenkins/plugins/git/GitStepTest.java (hunk 2 above)
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitStepTest.java]: When it was originally written…

**Comment by jglick** [comment]
Build timeout…

File: pom.xml [hunk 3]
```diff
@@ -28,7 +28,7 @@
     <concurrency>2</concurrency>
     <!-- TODO: Ongoing work on https://github.com/MarkEWaite/git-plugin/tree/master-findbugs-fixes. -->
     <findbugs.failOnError>false</findbugs.failOnError>
-    <workflow.version>1.14.2-SNAPSHOT</workflow.version> <!-- TODO https://github.com/jenkinsci/pipeline-plugin/pull/389 -->
+    <workflow.version>1.14.2</workflow.version>
```
Comment (MarkEWaite) [inline pom.xml]: How can I find the jenkins.version dependency of workflow.version 1.14.2?  I can't seem to find the 1.14.2 release of workflow-scm-step-plugin, and thus can't read its pom.

If this change will make the dependency 1.625 or newer, then I'll need to wait to merge it to the 3.0.0-beta branch.  If the change remains at 1.609 or earlier, then I can merge it to the master branch.

**Comment by MarkEWaite** [comment]
When I built this pull request and tried to run it in my docker Jenkins environment, it reported:

> The following plugins are deactivated because of cyclic dependencies, most likely you can resolve the issue by updating these to a newer version.
> workflow-scm-step
> git

Will there need to be a simultaneous release of workflow-scm-step and git plugins if this is merged to the git plugin master branch?

File: pom.xml (hunk 3 above)
Comment (jglick) [inline pom.xml]: It is 1.609.3. [Reference](https://github.com/jenkinsci/pipeline-plugin/blob/workflow-1.14.2/pom.xml#L52)

**Comment by jglick** [comment]
> The following plugins are deactivated because of cyclic dependencies

Hmm, what was the exact message, or way to reproduce? The whole reason for the 1.14.2 release was to avoid this problem, by making `workflow-scm-step` not depend on `git` any more. See comment earlier in the PR.

**Comment by MarkEWaite** [comment]
A docker instance that shows the problem is defined in my [lts-with-plugins-pr406](https://github.com/MarkEWaite/docker/tree/lts-with-plugins-pr406) branch.  When I build that docker image, then run it, the following text is included in the log output:

> Jun 06, 2016 1:09:03 PM hudson.PluginManager$1$3$2$1 reactOnCycle
> SEVERE: found cycle in plugin dependencies: (root=Plugin:workflow-scm-step, deactivating all involved) Plugin:workflow-scm-step -> Plugin:git -> Plugin:workflow-scm-step

**Comment by jglick** [comment]
I see, you are using `workflow-scm-step` 2.0 in that image, since your baseline is above 1.642.x. So I may need to forward-port https://github.com/jenkinsci/pipeline-plugin/pull/389 to `workflow-scm-step-plugin`, cut a release, then make https://github.com/jenkinsci/workflow-scm-step-plugin/issues/2 do an ours-merge of that.

**Comment by MarkEWaite** [comment]
Thanks for the clarification.  I have found that docker image to be quite useful, along with the 5-6 derivatives of it which I use in various scenarios.  I'll create another docker image based on the oldest Jenkins LTS currently of interest to me (1.609) and likely another on the first JDK 7 LTS (1.625).

I'd been ignoring those while I assembled a docker image with jobs that were interesting to me.

**Comment by jglick** [comment]
Prepared first step as https://github.com/jenkinsci/workflow-scm-step-plugin/pull/3.

**Comment by jglick** [comment]
I cut `workflow-scm-step` 2.1. It will probably take a few hours to make its way to the update center. Can you retry the integration test once that appears please?

**Comment by MarkEWaite** [comment]
I successfully used the docker Jenkins instance with 1.652 on my large scale private instance and didn't detect any surprises.

I was surprised when I merged a build of this pull request onto the git plugin master branch and  Jenkins 1.651.2 as represented in my [docker repository|https://github.com/MarkEWaite/docker/tree/16713a1a370a2dd7589a092c273e11a6b2e7e38d] on the [lts-with-plugins-pr406|https://github.com/MarkEWaite/docker/tree/lts-with-plugins-pr406] branch.

**Comment by jglick** [comment]
That warning is intentional: the `GitStep` in this plugin overrides the stub implementation, which is deleted in the downstream PR that cannot be merged until this PR (and another like it for Subversion) are released.

Make sense now? An unusually tricky change, I know.

**Comment by MarkEWaite** [comment]
Thanks for the explanation.  I've configured a 1.609 LTS docker instance and would like to run this change through that instance before I merge it.

**Pull request merged**